aten/src/ATen/core/TensorImpl.h

```diff
@@ -863,6 +874,54 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
    */
   bool is_variable() const { return is_variable_; };
 
+  /**
+   * True if a tensor is defined.
+   */
+  virtual bool defined() const {
+    return true;
+  }
```

Comment:
The segfault in ONNX export code in JIT tests is because we are trying to access the AutogradMeta pointer of a Variable that contains `UndefinedTensorImpl` in its `data_.impl_`. 

Ideally, the `defined()` function in Variable should check whether its `data_.impl_` is an `UndefinedTensorImpl`, and return false if it is. Right now we achieve this by making Tensor's `defined()` function also call `impl_->defined()`, and in Variable::Impl override the `defined()` function to check `data_.impl_->defined()`.

Reply:
I'm not really sure how we got there in the first place (a Variable where data_.impl_ is Undefined).  Can you explain?

Reply:
After a long time of digging I found the root cause: this line https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/python_ir.cpp#L360 does `std::move(v.data())`, which leaves the `data_` field empty for `v`, and hence we have an UndefinedTensorImpl in `v.data().impl_`. Changing `std::move(v.data())` to `v.data()` in that line fixes the issue.